File: scripts/hub/merge_upstream.py
#!/usr/bin/env python3
"""Download and merge upstream Surge modules into one bundle (deduped, sanitized)."""

from __future__ import annotations

import os
from datetime import datetime
from typing import Dict, List, Optional, Sequence, Tuple

from hub.common import Logger, read_file, safe_download
from hub.module_sanitizer import (
    format_header,
    format_module,
    merge_mitm_hosts,
    parse_module,
)

SourceSpec = Tuple[str, str]  # (label, url)


def _fetch_module(
    url: str,
    label: str,
    *,
    local_fallback: Optional[str] = None,
) -> str:
    if url.startswith(("http://", "https://")):
        content = safe_download(url, retries=2, timeout=60)
        if content and len(content.strip()) >= 20:
            return content
    elif os.path.isfile(url):
        text = "".join(read_file(url))
        if len(text.strip()) >= 20:
            return text
    if local_fallback and os.path.isfile(local_fallback):
        Logger.warn(f"Using local fallback for {label}")
        return "".join(read_file(local_fallback))
    raise RuntimeError(f"Failed to load upstream module: {label} ({url})")


def _combine_metadata(
    parts: Sequence[Tuple[str, Dict[str, str]]],
) -> Tuple[Optional[str], Optional[str], Optional[str]]:
    arg_tokens: List[str] = []
    args_desc_blocks: List[str] = []
    desc_blocks: List[str] = []

    for label, meta in parts:
        mod_name = meta.get("name", label).strip()

        # Handle arguments
        raw_args = meta.get("arguments", "").strip()
        if raw_args:
            for token in raw_args.split(","):
                token = token.strip()
                if token and token not in arg_tokens:
                    arg_tokens.append(token)

        # Handle arguments-desc
        raw_args_desc = meta.get("arguments-desc", "").strip()
        if raw_args_desc:
            import re
            clean_desc = re.sub(r'\n+', r'\\n', raw_args_desc.strip())
            args_desc_blocks.append(f"[{mod_name}]\\n{clean_desc}")

        # Handle desc
        raw_desc = meta.get("desc", "").strip()
        if raw_desc:
            import re
            clean_desc = re.sub(r'\n+', r'\\n', raw_desc.strip())
            desc_blocks.append(f"✦ {mod_name}: {clean_desc}")

    args_line = ",".join(arg_tokens) if arg_tokens else None
    
    def _beautify_and_truncate(blocks: List[str], max_len: int = 3800, prefix: str = "") -> Optional[str]:
        """Beautify and truncate description blocks to prevent Surge metadata overflow.
        
        Hard cap at ~3800 chars to stay safely under Surge's 4096 limit.
        """
        if not blocks:
            return None
        combined = prefix + "\\n\\n".join(blocks)
        if len(combined) > max_len:
            return combined[:max_len] + "…\\n(see upstream docs for full details)"
        return combined

    args_desc_line = _beautify_and_truncate(args_desc_blocks, 3800)
    desc_line = _beautify_and_truncate(desc_blocks, 3800, prefix="【Bundled Modules】\\n")
    return args_line, args_desc_line, desc_line


def merge_upstream_modules(
    sources: Sequence[SourceSpec],
    output_path: str,
    *,
    header_meta: Dict[str, str],
    provenance_comment: Optional[str] = None,
    optional_labels: Optional[Sequence[str]] = None,
    local_fallbacks: Optional[Dict[str, str]] = None,
    content_replacements: Optional[List[Tuple[str, str]]] = None,
) -> None:
    """Merge remote modules into output_path using module_sanitizer."""
    optional = set(optional_labels or ())
    fallbacks = local_fallbacks or {}
    parsed_parts: List[Tuple[str, Dict[str, str], List]] = []
    merged_sections: Dict[str, List[str]] = {}
    used_sources: List[SourceSpec] = []

    for label, url in sources:
        Logger.info(f"Downloading {label}...")
        try:
            text = _fetch_module(url, label, local_fallback=fallbacks.get(label))
            if content_replacements:
                for old_str, new_str in content_replacements:
                    text = text.replace(old_str, new_str)
        except RuntimeError as exc:
            if label in optional:
                Logger.warn(f"Skipping optional source {label}: {exc}")
                continue
            raise
        meta, sections = parse_module(text)
        parsed_parts.append((label, meta, sections))
        used_sources.append((label, url))
        for name, lines in sections:
            filtered_lines = [line for line in lines if line.strip()]
            if name in merged_sections and merged_sections[name] and filtered_lines:
                if merged_sections[name][-1] != "":
                    merged_sections[name].append("")
            merged_sections.setdefault(name, []).extend(filtered_lines)

    if not parsed_parts:
        raise RuntimeError(f"No upstream modules downloaded for {output_path}")

    combined_args, combined_args_desc, combined_desc = _combine_metadata(
        [(label, meta) for label, meta, _ in parsed_parts]
    )

    out_meta = dict(header_meta)
    out_meta["date"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    if combined_args:
        out_meta["arguments"] = combined_args
    if combined_args_desc:
        out_meta["arguments-desc"] = combined_args_desc
    if combined_desc:
        existing_desc = out_meta.get("desc", "").strip()
        if existing_desc:
            out_meta["desc"] = existing_desc + "\\n\\n" + combined_desc
        else:
            out_meta["desc"] = combined_desc

    section_list = [(name, merged_sections[name]) for name in merged_sections]
    section_list = merge_mitm_hosts(section_list)

    extra: List[str] = []
    if provenance_comment:
        extra.append(provenance_comment)
    extra.append("# Upstream modules merged by scripts/hub/merge_upstream.py")
    for label, url in used_sources:
        extra.append(f"# - {label}: {url}")

    header_lines = format_header(out_meta, extra_lines=extra)
    os.makedirs(os.path.dirname(output_path) or ".", exist_ok=True)
    with open(output_path, "w", encoding="utf-8") as f:
        f.write(format_module(header_lines, section_list, dedupe=True))

    _validate_merged_module(output_path)
    Logger.success(f"Wrote bundle: {output_path}")


def _validate_merged_module(output_path: str) -> None:
    """Reject merged bundles with known Surge-breaking merge artifacts."""
    text = "".join(read_file(output_path))
    if "%INSERT%" in text:
        raise RuntimeError(
            f"Merged module contains invalid MITM token %INSERT%: {output_path}"
        )
    if "hostname = %APPEND% %INSERT%" in text or "hostname = %INSERT%" in text:
        raise RuntimeError(f"Merged MITM hostname line is invalid: {output_path}")

File: go_scripts/pkg/hub/merge_upstream.go
package hub

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

type SourceSpec struct {
	Label string
	URL   string
}

func fetchModule(urlStr, label, localFallback, customUA string) (string, error) {
	if strings.HasPrefix(urlStr, "http://") || strings.HasPrefix(urlStr, "https://") {
		content := string(CurlFetch(urlStr, customUA))
		if content != "" && len(strings.TrimSpace(content)) >= 20 {
			time.Sleep(3 * time.Second) // strict delay to avoid shadowban/rate limits
			return content, nil
		}
	} else if ValidateFileExists(urlStr, "") {
		content := ReadFileString(urlStr)
		if len(strings.TrimSpace(content)) >= 20 {
			return content, nil
		}
	}

	if localFallback != "" && ValidateFileExists(localFallback, "") {
		Warn(fmt.Sprintf("Using local fallback for %s", label))
		return ReadFileString(localFallback), nil
	}

	return "", fmt.Errorf("failed to load upstream module: %s (%s)", label, urlStr)
}

func combineMetadata(parts []map[string]interface{}) (string, string, string) {
	var argTokens []string
	var argsDescBlocks []string
	var descBlocks []string

	for _, p := range parts {
		meta := p["meta"].(map[string]string)
		label := p["label"].(string)
		modName := meta["name"]
		if modName == "" {
			modName = label
		}

		rawArgs := strings.TrimSpace(meta["arguments"])
		if rawArgs != "" {
			for _, token := range strings.Split(rawArgs, ",") {
				token = strings.TrimSpace(token)
				if token != "" {
					found := false
					for _, t := range argTokens {
						if t == token {
							found = true
							break
						}
					}
					if !found {
						argTokens = append(argTokens, token)
					}
				}
			}
		}

		rawArgsDesc := strings.TrimSpace(meta["arguments-desc"])
		if rawArgsDesc != "" {
			cleanDesc := regexp.MustCompile(`\n+`).ReplaceAllString(rawArgsDesc, `\n`)
			argsDescBlocks = append(argsDescBlocks, fmt.Sprintf("[%s]\\n%s", modName, cleanDesc))
		}

		rawDesc := strings.TrimSpace(meta["desc"])
		if rawDesc != "" {
			cleanDesc := regexp.MustCompile(`\n+`).ReplaceAllString(rawDesc, `\n`)
			descBlocks = append(descBlocks, fmt.Sprintf("✦ %s: %s", modName, cleanDesc))
		}
	}

	argsLine := strings.Join(argTokens, ",")

	beautifyAndTruncate := func(blocks []string, maxLen int, prefix string) string {
		if len(blocks) == 0 {
			return ""
		}
		combined := prefix + strings.Join(blocks, "\\n\\n")
		if len(combined) > maxLen {
			return combined[:maxLen] + "…\\n(see upstream docs for full details)"
		}
		return combined
	}

	argsDescLine := beautifyAndTruncate(argsDescBlocks, 3800, "")
	descLine := beautifyAndTruncate(descBlocks, 3800, "【Bundled Modules】\\n")

	return argsLine, argsDescLine, descLine
}

func MergeUpstreamModules(sources []SourceSpec, outputPath string, headerMeta map[string]string, provenanceComment string, optionalLabels []string, localFallbacks map[string]string, contentReplacements [][2]string) error {
	optional := make(map[string]bool)
	for _, l := range optionalLabels {
		optional[l] = true
	}

	var parsedParts []map[string]interface{}
	mergedSections := make(map[string][]string)
	var usedSources []SourceSpec

	for _, src := range sources {
		Info(fmt.Sprintf("Downloading %s...", src.Label))

		customUA := ""
		if strings.Contains(src.URL, "yfamilys.com") {
			customUA = "Loon/3.9.9 CFNetwork/1496.0.7 Darwin/23.5.0"
		}

		fallback := ""
		if localFallbacks != nil {
			fallback = localFallbacks[src.Label]
		}

		text, err := fetchModule(src.URL, src.Label, fallback, customUA)
		if err != nil {
			if optional[src.Label] {
				Warn(fmt.Sprintf("Skipping optional source %s: %v", src.Label, err))
				continue
			}
			return err
		}

		if contentReplacements != nil {
			for _, repl := range contentReplacements {
				text = strings.ReplaceAll(text, repl[0], repl[1])
			}
		}

		meta, sections := ParseModule(text)
		parsedParts = append(parsedParts, map[string]interface{}{
			"label":    src.Label,
			"meta":     meta,
			"sections": sections,
		})
		usedSources = append(usedSources, src)

		for _, sec := range sections {
			name := sec.Name
			var filteredLines []string
			for _, line := range sec.Lines {
				if strings.TrimSpace(line) != "" {
					filteredLines = append(filteredLines, line)
				}
			}

			if len(mergedSections[name]) > 0 && len(filteredLines) > 0 {
				if mergedSections[name][len(mergedSections[name])-1] != "" {
					mergedSections[name] = append(mergedSections[name], "")
				}
			}
			mergedSections[name] = append(mergedSections[name], filteredLines...)
		}
	}

	if len(parsedParts) == 0 {
		return fmt.Errorf("no upstream modules downloaded for %s", outputPath)
	}

	combinedArgs, combinedArgsDesc, combinedDesc := combineMetadata(parsedParts)

	outMeta := make(map[string]string)
	for k, v := range headerMeta {
		outMeta[k] = v
	}

	outMeta["date"] = time.Now().Format("2006-01-02 15:04:05")
	if combinedArgs != "" {
		outMeta["arguments"] = combinedArgs
	}
	if combinedArgsDesc != "" {
		outMeta["arguments-desc"] = combinedArgsDesc
	}
	if combinedDesc != "" {
		existing := strings.TrimSpace(outMeta["desc"])
		if existing != "" {
			outMeta["desc"] = existing + "\\n\\n" + combinedDesc
		} else {
			outMeta["desc"] = combinedDesc
		}
	}

	var sectionList []ModuleSection
	for name, lines := range mergedSections {
		sectionList = append(sectionList, ModuleSection{Name: name, Lines: lines})
	}

	sectionList = MergeMitmHosts(sectionList)

	var extra []string
	if provenanceComment != "" {
		extra = append(extra, provenanceComment)
	}
	extra = append(extra, "# Upstream modules merged by go_scripts/pkg/pipeline/merge_bundles.go")
	for _, src := range usedSources {
		extra = append(extra, fmt.Sprintf("# - %s: %s", src.Label, src.URL))
	}

	headerLines := FormatHeader(outMeta, extra)
	EnsureDir(filepath.Dir(outputPath))

	content := FormatModule(headerLines, sectionList, true, nil)
	if err := SafeWriteFile(outputPath, content, true); err != nil {
		return err
	}

	if strings.Contains(content, "%INSERT%") {
		return fmt.Errorf("merged module contains invalid MITM token %%INSERT%%: %s", outputPath)
	}

	Success(fmt.Sprintf("Wrote bundle: %s", outputPath))
	return nil
}
